### PR TITLE
Add wlan interface to domd

### PR DESCRIPTION
**Pull request by dsemenets**

### CODE REVIEW

File: recipes-kernel/linux/files/bcm2712-raspberrypi5-domd.dts [hunk 1]
```diff
@@ -34,13 +34,72 @@
 			sdhci-caps = <0x0 0x0>;
 			sdhci-caps-mask = <0xc000 0x0>;
 			xen,path = "/axi/mmc@fff000";
-			xen,force-assign-without-iommu = <0x1>;
+			xen,force-assign-without-iommu;
```
Comment (lorc) [inline recipes-kernel/linux/files/bcm2712-raspberrypi5-domd.dts]: This is the correct change, but it should be in the separate commit, as it merely fixes prior mistake.

**Comment by GrygiriiS** [comment]
Sorry, but It is not how it should be - we can't blindly and unconditionally add devices in Dom1 (DomD).
For example, now we pass MMC and Sd-card is used for rootfs, but our target is USB and USB stick with rootfs.
Networking, 4 cases possible: no net; Ethernet; wifi;  Ethernet and wifi. And it should be configurable. 

So there should be build options and DT split somehow (probably overlays). So final DT and rootfs are built based on configuration options..

**Comment by lorc** [comment]
NAK. Please read https://xen-troops.github.io/process/process

`linux: domd.dts: Minor fixes.` is not a valid commit description

**Comment by oleksiimoisieiev** [comment]
@dsemenets Are you going to fix comments from colleagues?

File: recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso [hunk 2]
```diff
@@ -0,0 +1,127 @@
+/*
```
Comment (oleksiimoisieiev) [inline recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso:2]: gio? Maybe gpio

File: recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+			pins_dat {
+				function = "sd2";
+				pins = "gpio32\0gpio33\0gpio34\0gpio35";
+				bias-pull-up;
+			};
+		};
+	};
+
+	mmc@1001100000 {
```
Comment (oleksiimoisieiev) [inline recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso:46]: do we need mmc for wifi?

File: recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb [hunk 4]
```diff
@@ -16,6 +16,7 @@ SRC_URI = " \
 "
 MMC_PASSTHROUGH_DTBO = "mmc-passthrough.dtbo"
 USB_PASSTHROUGH_DTBO = "usb-passthrough.dtbo"
+WIFI_PASSTHROUGH_DTBO = "wifi-passthrough.dtbo"
```
Comment (oleksiimoisieiev) [inline recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: I don't think we need a separate dtbo. It should be configurable

File: recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso (hunk 2 above)
Comment (dsemenets) [reply recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso:2]: No. gpio7d50800 is gio in rpi5 dt

File: recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso (hunk 3 above)
Comment (dsemenets) [reply recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso:46]: wifi is child for this node

File: recipes-kernel/linux/files/bcm2712-raspberrypi5-domd.dts (hunk 1 above)
Comment (dsemenets) [reply recipes-kernel/linux/files/bcm2712-raspberrypi5-domd.dts]: Moved to separate PR

File: recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb (hunk 4 above)
Comment (dsemenets) [reply recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: This dtbo applied if MACHINE_FEATURES contains domd_wifi

File: recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso [hunk 5]
```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2024 EPAM systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+*/
+/dts-v1/;
+/plugin/;
+
+&passthrough {
+	pinctrl@107d504100 {
```
Comment (GrygiriiS) [inline recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso:11]: pinctrl looks like a problem, unfortunately :( We can't jsut passthrough it, as it's shared resource. All current HW working with assumption pinctrl is configured already by RPI5 bootloader or u-boot.

Comment (GrygiriiS) [reply recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso:11]: UPD: Not sure how to proceed :(

File: recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso (hunk 2 above)
Comment (GrygiriiS) [reply recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso:2]: Yep. it's gio

File: recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso (hunk 5 above)
Comment (lorc) [reply recipes-kernel/linux/files/bcm2712-raspberrypi5-wifi.dtso:11]: Yeah, in cases like this we are considering using SCMI. This is the best approach, but it is PITA to implement. 

Another approach is to configure pinctrl in bootloader/dom0 so guest can use static configuration. Of course, it will not work if dynamic configuration is required.

File: recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb (hunk 4 above)
Comment (oleksiimoisieiev) [reply recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: I understand. but I see here that we have two configurations:
domd via usb, domd via mmc and domd via wifi but usb and mmc disabled. Is it true?
I see wifi as a optional feature which either can be enabled or disabled. Or if it's mandatory or it is a problem to make it optional then don't add this option

Comment (dsemenets) [reply recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: It is optional feature. It can be used with rootfs on usb or mmc. Overlays are not applied if MACHINE_FEATURES doesn't contain domd_wifi

Comment (oleksiimoisieiev) [reply recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: So why do we have 3 dtbo ? for mmc, usb and wifi?

Comment (GrygiriiS) [reply recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: The overlays applied to add features: 
```
usb-passthrough.dtbo = USB + NET
usb-passthrough.dtbo + wifi-passthrough.dtbo = USB + NET + WIFI
mmc-passthrough.dtbo + wifi-passthrough.dtbo = MMC + NET + WIFI
mmc-passthrough.dtbo + wifi-passthrough.dtbo + usb-passthrough.dtbo = MMC + USB +NET + WIFI

... etc
```

Comment (dsemenets) [reply recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: Yes. But combination mmc-passthrough.dtbo + wifi-passthrough.dtbo + usb-passthrough.dtbo wasn't tested

Comment (GrygiriiS) [reply recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: Of course not, it was a generic answer on generic question "Why we need  dtbo"

Comment (dsemenets) [reply recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: > So why do we have 3 dtbo ? for mmc, usb and wifi?

Yes

Comment (oleksiimoisieiev) [reply recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:22]: Ok thanks

**Comment by oleksiimoisieiev** [comment]
LGTM `Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

**Comment by lorc** [comment]
We are passing SoC pinctrl to DomD, which is not secure, as DomD may (and probably will) mess with SDHC used in Dom0.

Therefore, it is NAK for current implementation. Please consider using SCMI or configuring pinctrl from a bootloader.

**Comment by oleksiimoisieiev** [comment]
> We are passing SoC pinctrl to DomD, which is not secure, as DomD may (and probably will) mess with SDHC used in Dom0.
> 
> Therefore, it is NAK for current implementation. Please consider using SCMI or configuring pinctrl from a bootloader.

I understand your concerns and agree with them. Looks like making pinctrl config on u-boot is the only option for us. That's because we don't have a full support for arm-trusted-firmware. This means that SCMI support implementation may require a lot of effort (unless I've missed something and rpi5 loader has scmi support).
So the only option I see for now is to make u-boot support. Especially counting the fact that u-boot already has pinctl driver.
@GrygiriiS @lorc what do you think?

**Comment by lorc** [comment]
> (unless I've missed something and rpi5 loader has scmi support).

No, but we have TF-A running, so we can use TF-A based SCMI. I believe, you are the author of these patches?

> So the only option I see for now is to make u-boot support. 

It makes the whole system depending on U-Boot which is already bad. But bigger problem is that we can't change pinctrl (and clocks and resets) dynamically.

**Comment by klogg** [comment]
@oleksiimoisieiev will take ownership of these patches

**Comment by oleksiimoisieiev** [comment]
> @oleksiimoisieiev will take ownership of these patches

Taking to work. waiting for SCMI to be finished.

**Comment by GrygiriiS** [comment]
Not needed any more. all required parts incorporated in
https://github.com/xen-troops/meta-xt-prod-devel-rpi5/pull/41
